Posts/2021/03/Caps-lock-ToUpper.md

```diff
@@ -0,0 +1,206 @@
+
+So, in general, you should never trust any user input without validating it first.
+Although you ask the user to type her name in all upper case, I'll bet that many just won't.
+
+So what does the Caps Lock actually do?
+When you type characters into a form or a console, you might type them like this:
+
+```powershell-console
+this is my sentence.
+```
+
+If you switch on the Caps Lock key, the operating system and your hardware makes those characters appear like this:
+
+```console
+THIS IS MY SENTENCE.
+```
+
+So how can we achieve that same effect in a script?
+Simple: we accept the input as the user typed it.
+Then we make sure it's all upper case before using it.
+
+Let's start with getting the user input in the first place.
+
+## Getting User Input
+
+There are several ways to get user input from within a script.
+A common approach with PowerShell scripts is to use the `Read-Host` command.
+This cmdlet reads a line of input from the console and returns it to the script as a string.
+For more information on this cmdlet, see the [Read-Host help file](https://docs.microsoft.com/powershell/module/microsoft.powershell.utility/read-host).
+
+There are other ways to get user input, such as using [Windows Forms](https://docs.microsoft.com/powershell/scripting/samples/creating-a-custom-input-box?view=powershell-7.1) or WPF.
+You might even use a legacy [Visual Basic `Inputbox`](https://docs.microsoft.com/dotnet/api/microsoft.visualbasic.interaction.inputbox?view=net-5.0).
+But with each of these methods, you still have the underlying issue of making sure the string the user enters is all upper-case before you use it further.
+
+Suppose you wanted to ask the user for their name (and you really need it to be upper case).
+You could ask for, accept, and then display user input like this:
+
+```powershell-console
+PS C:\Foo> $Answer = Read-Host -Prompt 'Please Enter Your Name In ALL Upper case'
+Please Enter Your Name In ALL Upper case: Thomas Lee
+PS C:\Foo> $Answer
+Thomas Lee
+```
+
+But that is not in upper case, I hear you say.
+Yes, true - but there is just one more step.
+Be patient, grasshopper.
+
+## Converting a String to Upper Case
+
+As I mentioned, when you use `Read-Host`, PowerShell returns the input to you as a string.
+Even if you enter a number (say 42) PowerShell still treats this as a string containing two characters, like this:
+
+```powershell-console
+PS C:\> $Answer = Read-Host -Prompt "Please Enter Your Name In ALL Upper case"
+Please Enter Your Name In ALL Upper case: 42
+PS C:\> $Answer.GetType().FullName
+System.String
+```
+
+This matters because the `System.String` .NET class has a very useful method, **ToUpper()**.
+The **ToUpper()** method converts the string to all upper case and returns a new, all upper case, string.
+So to convert the string you entered and stored in **$Answer**, you use the **ToUpper()** method like this:
```

Comment:
```suggestion
This matters because the `System.String` .NET class has a very useful method, `ToUpper()`.
The `ToUpper()` method converts the string to all upper case and returns a new, all upper case, string.
So to convert the string you entered and stored in `$Answer`, you use the `ToUpper()` method like this:
```